new/ohai-config.md

```diff
@@ -0,0 +1,61 @@
+---
+RFC: unassigned
+Author: Bryan McLellan <btm@loftninjas.org>
+Status: Draft
+Type: Standards Track
+---
+
+# Ohai User Configuration
+
+Provides a standard for passing configuration settings to Ohai to configure its behavior as well as that of plugins.
+
+## Motivation
+
+    As a Chef user,
+    I want to easily configure Ohai,
+    so that it behaves optimally on my infrastructure.
+
+    As an Ohai user,
+    I want Ohai to load configuration settings from my client.rb,
+    so that it behaves the same as it does during a chef-client run.
+
+## Specification
+
+### Configuration File
+
+Previously Ohai has not had a configuration file. A small number of Ohai configuration settings do exist that can be specified in the chef-client ```client.rb``` file, which are then passed to Ohai during a chef-client run. For example, additional locations on the file system can be searched for plugins by adding paths to the ```Ohai::Config[:plugin_path]``` array. This file is only loaded by Chef and does not affect Ohai when being run on the command line nor any other programs loading Ohai as a library.
+
+When run from the command line, Ohai should load the ```client.rb``` file from the appropriate platform specific path unless an alternate configuration file is provided as a command line argument. To preserve compatibility with the existing parsing of the ```client.rb```, the Ohai configuration file syntax will be variables set on the ```Ohai::Config``` class,, and barewords will be ignored using method_missing.
+
+For example:
+
+```
+Ohai::Config[:plugin_path] << "/usr/local/lib/ohai/plugins"
+Ohai::Config[:plugin][:hostname][:fqdn_using] = [ :hostname, :nis, :dns ]
+
+# This is ignored:
+node_name "darius"
+```
+
+Using the same ```client.rb``` file used by the chef-client also allows both tools to be configured simultaneously by populating this file during bootstrap with custom local values (when that is supported in the future).
+
+### Namespacing
+
+To reduce the risk of built-in and custom plugins using the same configuration setting for conflicting purposes, plugins must prefix their configuration settings with ```[:plugin]``` and the name of the plugin consuming the setting. 
+
+For example:
+
+```
+Ohai::Config[:plugin][:dmi][:all_ids] = true
+Ohai::Config[:plugin][:ec2][:silly_magic_arp] = "de:ad:de:ad:de:ad"
+Ohai::Config[:plugin][:platform][:amazon_is_amazon] = true
```

Comment:
Yeah. Mixlib-config is case sensitive. I would guess it's technically possibly to have a ```ec2``` and an ```EC2``` named Ohai plugin, but that would be maddening. I'd support codifying in this RFC that all key names should be lower-case.

There's a fine point here as to if we're using the filename of the plugin or the plugin name. The passwd.rb plugin contains this line ```Ohai.plugin(:Passwd) do```, which is the plugin name as @mcquin points out.

There are nine ```:Filesystem``` plugins in different paths like ```windows/filesystem.rb``` and ```aix/filesystem.rb```. In that cause you'd have this:

```
Ohai::Config[:plugin][:windows][:filesystem][:ignore_drives] = [ "z", "x" ]
Ohai::Config[:plugin][:aix][:filesystem][:whatever_aix_does] = "banana ice cream"
```

Here's the thing, we're not enforcing this with code. We're not going to have mixlib-config prohibit you from setting values in a portion of the config that's different than your plugin name or plugin path. I think we're only looking for something that's 1) simple and 2) likely to reduce name collisions.

More often than not the name of the file on disk is the name of the plugin. Right now I'm planning on changing my proposal to say you should name space as the lowercase version of the plugin name. People should generally get that right. In places where there's a problem like ```darwin/system_profiler.rb``` is SystemProfile, the correct namespace would be Ohai::Config[:plugin][:systemprofile]. We can catch people choosing the wrong one when we review code, but it's not a catastrophe if they don't